docs/src/extend/custom-rule-tutorial.md

```diff
@@ -0,0 +1,452 @@
+
+```javascript
+// enforce-foo-bar.js
+
+module.exports = {
+    meta: {
+        type: "problem",
+        docs: {
+            description: "Enforce that a variable named `foo` can only be assigned a value of 'bar'.",
+        },
+        fixable: "code"
+    },
+    create(context) {
+        return {
+            // TODO: add callback function(s)
+            }
+        };
+    }
+};
+```
+
+To learn more about rule metadata, refer to [Rule Structure](custom-rules#rule-structure).
+
+## Step 4: Add Rule Visitor Methods
+
+Define the rule's `create` function, which accepts a `context` object and returns an object with a property for each syntax node type you want to handle. In this case, you want to handle `VariableDeclaration` nodes.
+You can choose any [ESTree node type](https://github.com/estree/estree) or [selector](selectors).
+
+Inside the `VariableDeclaration` visitor method, check if the node represents a `const` variable declaration, if its name is `foo`, and if it's not assigned to the string `"bar"`. You do this by evaluating the `node` passed to the `VariableDeclaration` method.
+
+If the `const foo` declaration is assigned a value of `"bar"`, then the rule does nothing. If `const foo` **is not** assigned a value of `"bar"`, then `context.report()` reports an error to ESLint. The error report includes information about the error and how to fix it.
+
+```javascript
+// enforce-foo-bar.js
+module.exports = {
+    meta: {
+        type: "problem",
+        docs: {
+            description: "Enforce that a variable named `foo` can only be assigned a value of 'bar'.",
+        },
+         fixable: "code"
+    },
+    create(context) {
+        return {
+            // Performs action in the function on every variable declaration
+            VariableDeclaration(node) {
+                // Check if a `const` variable declaration
+                if(node.kind === "const") {
+                    // Check if variable name is `foo`
+                    if(node.declarations[0].id.name === "foo") {
+                        // Check if the value of the variable is "bar"
+                        if (node.declaration[0].init.value !== "bar") {
+                            // Report error to ESLint. Error message uses
+                            // a message placeholder to include the incorrect value
+                            // in the error message.
+                            // Also includes a `fix(fixer)` function that replaces
+                            // any values assigned to `const foo` with "bar".
+                            context.report({
+                                node,
+                                message: 'Value other than "bar" assigned to \`const foo\`. Unexpected value: {{ notBar }}',
+                                data: {
+                                    notBar: node.declaration.init.value
```

Comment:
```suggestion
                                    notBar: node.declarations[0].init.value
```

Also, it looks like this assumes that `node.declarations[0].init` is a Literal node. Should the rule catch only literals (in which case we should update the rule code) or any nodes (in which case we should update this line)?

Reply:
to handle other types of nodes, what would the code update be here? and what other types of non-literal nodes would make sense to test?

Reply:
toward the top of the text, updating to be '"assigned to the string **literal** `"bar"`':

```md
## The Custom Rule

The custom rule  in this tutorial requires that all `const` variables named `foo` are assigned the string literal `"bar"`. The rule is defined in the file `enforce-foo-bar.js`. The rule also suggests replacing any other value assigned to `const foo` with `"bar"`.
```

i'd rather not get to deep into this topic since i think it'd be a distraction from the beginner-oriented tutorial